File: OnlineStore/Ecommerce.Sdk.Controls/WishList/WishList.cs
/**
 * SAMPLE CODE NOTICE
 * 
 * THIS SAMPLE CODE IS MADE AVAILABLE AS IS.  MICROSOFT MAKES NO WARRANTIES, WHETHER EXPRESS OR IMPLIED,
 * OF FITNESS FOR A PARTICULAR PURPOSE, OF ACCURACY OR COMPLETENESS OF RESPONSES, OF RESULTS, OR CONDITIONS OF MERCHANTABILITY.
 * THE ENTIRE RISK OF THE USE OR THE RESULTS FROM THE USE OF THIS SAMPLE CODE REMAINS WITH THE USER.
 * NO TECHNICAL SUPPORT IS PROVIDED.  YOU MAY NOT DISTRIBUTE THIS CODE UNLESS YOU HAVE A LICENSE AGREEMENT WITH MICROSOFT THAT ALLOWS YOU TO DO SO.
 */

using System;
using System.Collections.ObjectModel;
using System.Runtime.InteropServices;
using System.Security.Permissions;
using System.Web;
using System.Web.UI;

[assembly: WebResource("Contoso.Retail.Ecommerce.Sdk.Controls.WishList.WishList.css", "text/css", PerformSubstitution = true)]

namespace Contoso
{
    namespace Retail.Ecommerce.Sdk.Controls
    {
        /// <summary>
        /// The wish list control.
        /// </summary>
        [AspNetHostingPermission(SecurityAction.Demand, Level = AspNetHostingPermissionLevel.Minimal)]
        [ToolboxData("<{0}:WishList runat=server></{0}:WishList>")]
        [ComVisible(false)]
        public sealed class WishList : RetailWebControl
        {
            /// <summary>
            /// Gets or sets a value indicating the amount of wish lists shown.
            /// </summary>
            public int WishListCount { get; set; }

            /// <summary>
            /// Gets or sets a value indicating whether the wish list is displayed in full view or condensed view.
            /// </summary>
            public bool IsFullView { get; set; }

            /// <summary>
            /// Gets or sets the Url of the page containing line by line details of a wish list.
            /// </summary>
            public string WishListDetailsUrl { get; set; }

            /// <summary>
            /// Gets the markup to include control scripts, CSS, startup scripts in the page.
            /// </summary>
            /// <returns>The header markup.</returns>
            internal new string GetHeaderMarkup()
            {
                Collection<string> cssUrls = this.GetCssUrls();
                Collection<string> scriptUrls = this.GetScriptUrls();

                string existingHeaderMarkup = this.GetExistingHeaderMarkup();
                string output = this.GetCssAndScriptMarkup(existingHeaderMarkup, cssUrls, scriptUrls);
                output += this.GetStartupMarkup(existingHeaderMarkup);

                return output;
            }

            /// <summary>
            /// Gets the control html markup with the control class wrapper added.
            /// </summary>
            /// <returns>The control markup.</returns>
            internal string GetControlMarkup()
            {
                return base.GetControlMarkup(WishList.CssClassName);
            }

            /// <summary>
            /// Raises the PreRender event.
            /// </summary>
            /// <param name="e">An <see cref="System.EventArgs"/> object that contains the event data.</param>
            protected override void OnPreRender(EventArgs e)
            {
                base.OnPreRender(e);

                string headerMarkup = base.GetHeaderMarkup();
                headerMarkup += this.GetHeaderMarkup();
                this.RegisterHeaderMarkup(headerMarkup);
            }

            /// <summary>
            /// Renders the contents of the control to the specified writer. This method is used primarily by control developers.
            /// </summary>
            /// <param name="writer">A <see cref="System.Web.UI.HtmlTextWriter"/> that represents the output stream to render HTML content on the client.</param>
            protected override void RenderContents(HtmlTextWriter writer)
            {
                base.RenderContents(writer);
            }

            /// <summary>
            /// Gets the control html markup.
            /// </summary>
            /// <returns>The control markup.</returns>
            protected override string GetHtml()
            {
                return this.GetHtmlFragment("WishList.WishList.html");
            }

            /// <summary>
            /// Gets the script URLs.
            /// </summary>
            /// <returns>The script URLs.</returns>
            private new Collection<string> GetScriptUrls()
            {
                return new Collection<string>();
            }

            /// <summary>
            /// Gets the CSS URLs.
            /// </summary>
            /// <returns>The CSS URLs.</returns>
            private new Collection<string> GetCssUrls()
            {
                Collection<string> cssUrls = new Collection<string>();
                cssUrls.Add(this.GetFileUrl("WishList.WishList.css"));

                return cssUrls;
            }

            /// <summary>
            /// Gets the startup markup.
            /// </summary>
            /// <param name="existingHeaderMarkup">Existing header markup to determine if startup script registration is required.</param>
            /// <returns>The startup markup.</returns>
            private new string GetStartupMarkup(string existingHeaderMarkup)
            {
                string startupScript = string.Empty;

                string wishListStartupScript = string.Format(
                    @"<script type='text/javascript'> 
                    msaxValues['msax_WishListCount'] = '{0}';
                    msaxValues['msax_WishListIsFullView'] = '{1}';
                    msaxValues['msax_WishListDetailsUrl'] = '{2}';
                </script>",
                    this.WishListCount,
                    this.IsFullView,
                    this.WishListDetailsUrl);

                // If page header is visible here, check if the markup is present in the header already.
                if (this.Page == null || this.Page.Header == null || (existingHeaderMarkup != null && !existingHeaderMarkup.Contains(wishListStartupScript)))
                {
                    startupScript += wishListStartupScript;
                }

                return startupScript;
            }
        }
    }
}

File: OnlineStore/Ecommerce.Sdk.Controls/WishList/WishList.ts
/**
 * SAMPLE CODE NOTICE
 * 
 * THIS SAMPLE CODE IS MADE AVAILABLE AS IS.  MICROSOFT MAKES NO WARRANTIES, WHETHER EXPRESS OR IMPLIED,
 * OF FITNESS FOR A PARTICULAR PURPOSE, OF ACCURACY OR COMPLETENESS OF RESPONSES, OF RESULTS, OR CONDITIONS OF MERCHANTABILITY.
 * THE ENTIRE RISK OF THE USE OR THE RESULTS FROM THE USE OF THIS SAMPLE CODE REMAINS WITH THE USER.
 * NO TECHNICAL SUPPORT IS PROVIDED.  YOU MAY NOT DISTRIBUTE THIS CODE UNLESS YOU HAVE A LICENSE AGREEMENT WITH MICROSOFT THAT ALLOWS YOU TO DO SO.
 */

/// <reference path="../Common/Helpers/Core.ts" />
/// <reference path="../Resources/Resources.ts" />

module Contoso.Retail.Ecommerce.Sdk.Controls {
    "use strict";

    export class WishList {

        public errorMessages: ObservableArray<string>;
        private wishListsExist;
        private wishLists: ObservableArray<CommerceProxy.Entities.CommerceList>;
        private wishListToBeUpdated: Observable<CommerceProxy.Entities.CommerceList>;
        private wishListToBeDeleted: Observable<CommerceProxy.Entities.CommerceList>;
        private _wishListView;
        private _loadingDialog;
        private _loadingText;
        private errorPanel;
        private deleteWishListConfirmationModal;
        private editWishListModal;
        private createWishListModal;
        private dialogOverlay;

        private wishListCount: number = parseInt(msaxValues.msax_WishListCount);
        private isFullView;

        constructor(element) {
            this._wishListView = $(element);
            this._loadingDialog = this._wishListView.find('.msax-Loading');
            this._loadingText = this._loadingDialog.find('.msax-LoadingText');
            this.errorPanel = this._wishListView.find(" > .msax-ErrorPanel");
            this.deleteWishListConfirmationModal = this._wishListView.find('.msax-DeleteWishListConfirmationModal');
            this.editWishListModal = this._wishListView.find('.msax-EditWishListModal');
            this.createWishListModal = this._wishListView.find('.msax-CreateWishListModal');
            LoadingOverlay.CreateLoadingDialog(this._loadingDialog, this._loadingText, 200, 200);
            this.errorMessages = ko.observableArray<string>([]);

            this.isFullView = ko.computed(() => {
                return (Utils.isNullOrUndefined(msaxValues.msax_WishListIsFullView) ? false : msaxValues.msax_WishListIsFullView.toLowerCase() == "true");
            });

            this.wishLists = ko.observableArray<CommerceProxy.Entities.CommerceList>([]);
            this.wishListToBeUpdated = ko.observable<CommerceProxy.Entities.CommerceList>(null);
            this.wishListToBeDeleted = ko.observable<CommerceProxy.Entities.CommerceList>(null);

            this.wishListsExist = ko.computed(() => {
                return (this.wishLists().length != 0);
            });

            this.getWishLists();
        }

        private getResx(key: string) {
            // Gets the resource value.
            return Resources[key];
        }

        private editWishListOverlayClick() {
            this.dialogOverlay = $('.ui-widget-overlay');
            this.dialogOverlay.on('click', $.proxy(this.closeEditWishListDialog, this));
        }

        private openDeleteWishListConfirmationDialog(args: any) {
            this.wishListToBeDeleted(args);
            if (!Utils.isNullOrEmpty(this.deleteWishListConfirmationModal)) {
                this.deleteWishListConfirmationModal.modal('show');
            }
        }

        private closeDeleteWishListConfirmationDialog() {
            if (!Utils.isNullOrEmpty(this.deleteWishListConfirmationModal)) {
                this.deleteWishListConfirmationModal.modal('hide');
            }
        }

        private openEditWishListDialog(args: any) {
            this.wishListToBeUpdated(args);
            if (!Utils.isNullOrEmpty(this.editWishListModal)) {
                this.editWishListModal.modal('show');
            }
        }

        private closeEditWishListDialog() {
            if (!Utils.isNullOrEmpty(this.editWishListModal)) {
                this.editWishListModal.modal('hide');
            }
        }

        private openCreateWishListDialog() {
            if (!Utils.isNullOrEmpty(this.createWishListModal)) {
                this.createWishListModal.modal('show');
            }
        }

        private closeCreateWishListDialog() {
            if (!Utils.isNullOrEmpty(this.createWishListModal)) {
                this.createWishListModal.modal('hide');
            }
        }

        private showError(isError: boolean) {
            // Shows the error message on the error panel.
            if (isError) {
                this.errorPanel.addClass("msax-Error");
            }
            else if (this.errorPanel.hasClass("msax-Error")) {
                this.errorPanel.removeClass("msax-Error");
            }

            this.errorPanel.show();
            $(window).scrollTop(0);
        }

        private getWishListDetailsUrl(wishList: CommerceProxy.Entities.CommerceList): string {
            var url: string = msaxValues.msax_WishListDetailsUrl;

            url += "?wishListId=" + wishList.Id;

            return url;
        }

        // Service calls
        private getWishLists() {
            CommerceProxy.RetailLogger.getWishListsStarted();
            LoadingOverlay.ShowLoadingDialog();
            CustomerWebApi.GetWishListsByCustomer(this)
                .done((responseWishLists: CommerceProxy.Entities.CommerceList[]) => {
                    this.wishLists(responseWishLists);
                    LoadingOverlay.CloseLoadingDialog();
                    CommerceProxy.RetailLogger.getWishListsFinished();
                })
                .fail((errors: CommerceProxy.ProxyError[]) => {
                    Core.LogEvent(CommerceProxy.RetailLogger.getWishListsError, errors, Resources.String_254); // Sorry, something went wrong. The wish list information couldn't be retrieved. Please refresh the page and try again.
                    this.errorMessages([Resources.String_254]); // Sorry, something went wrong. The wish list information couldn't be retrieved. Please refresh the page and try again.
                    this.showError(true);
                    LoadingOverlay.CloseLoadingDialog();
                });
        }

        private createWishList() {
            CommerceProxy.RetailLogger.createWishListStarted();
            LoadingOverlay.ShowLoadingDialog();
            var newWishListName: string = this.createWishListModal.find('#CreateWishListTextBox').val();

            WishListWebApi.CreateWishList(newWishListName, this)
                .done((wishList: CommerceProxy.Entities.CommerceList) => {
                    if (!Utils.isNullOrUndefined(wishList)) {
                        this.closeCreateWishListDialog();
                    } else {
                        this.showError(true);
                    }

                    this.getWishLists();
                    LoadingOverlay.CloseLoadingDialog();
                    CommerceProxy.RetailLogger.createWishListFinished();
                })
                .fail((errors: CommerceProxy.ProxyError[]) => {
                    Core.LogEvent(CommerceProxy.RetailLogger.createWishListError, errors, Resources.String_251); // Sorry, something went wrong. We couldn't create the wish list.
                    var errorMessages: string[] = ErrorHelper.getErrorMessages(errors);
                    this.showError(true);
                    LoadingOverlay.CloseLoadingDialog();
                });
        }

        private deleteWishList() {
            CommerceProxy.RetailLogger.deleteWishListStarted();
            LoadingOverlay.ShowLoadingDialog();
            var deletedWishList: CommerceProxy.Entities.CommerceList = this.wishListToBeDeleted();

            WishListWebApi.DeleteWishList(deletedWishList.Id.toString(), this)
                .done((data: CommerceProxy.Entities.CommerceList) => {
                    this.getWishLists();
                    this.deleteWishListConfirmationModal.modal('hide');
                    LoadingOverlay.CloseLoadingDialog();
                    CommerceProxy.RetailLogger.deleteWishListFinished();
                })
                .fail((errors: CommerceProxy.ProxyError[]) => {
                    Core.LogEvent(CommerceProxy.RetailLogger.deleteWishListError, errors, Resources.String_252); // Sorry, something went wrong. We couldn't delete the wish list.
                    var errorMessages: string[] = ErrorHelper.getErrorMessages(errors);
                    this.showError(true);
                    LoadingOverlay.CloseLoadingDialog();
                });
        }

        private updateWishList() {
            CommerceProxy.RetailLogger.updateWishListStarted();
            LoadingOverlay.ShowLoadingDialog();
            var newWishListName: string = this.editWishListModal.find('#EditWishListTextBox').val();
            var updatedWishList: CommerceProxy.Entities.CommerceList = this.wishListToBeUpdated();
            updatedWishList.Name = newWishListName;

            WishListWebApi.UpdateWishList(updatedWishList, this)
                .done((wishList: CommerceProxy.Entities.CommerceList) => {
                    if (!Utils.isNullOrUndefined(wishList)) {
                        this.closeEditWishListDialog();
                    } else {
                        this.showError(true);
                    }

                    this.getWishLists();
                    LoadingOverlay.CloseLoadingDialog();
                    CommerceProxy.RetailLogger.updateWishListFinished();
                })
                .fail((errors: CommerceProxy.ProxyError[]) => {
                    Core.LogEvent(CommerceProxy.RetailLogger.updateWishListError, errors, Resources.String_253); // Sorry, something went wrong. We couldn't edit the wish list details. Please refresh the page and try again.
                    var errorMessages: string[] = ErrorHelper.getErrorMessages(errors);
                    this.showError(true);
                    LoadingOverlay.CloseLoadingDialog();
                });
        }

        private markWishListAsFavorite(args: CommerceProxy.Entities.CommerceList) {
            CommerceProxy.RetailLogger.updateWishListStarted();
            LoadingOverlay.ShowLoadingDialog(Utils.format("The wish list {0} is your new favorite", args.Name));
            this.wishListToBeUpdated(args);
            var updatedWishList: CommerceProxy.Entities.CommerceList = this.wishListToBeUpdated();
            updatedWishList.IsFavorite = true;
            WishListWebApi.UpdateWishList(updatedWishList, this)
                .done((wishList: CommerceProxy.Entities.CommerceList) => {
                    this.getWishLists();
                    LoadingOverlay.CloseLoadingDialog();
                    CommerceProxy.RetailLogger.updateWishListFinished();
                })
                .fail((errors: CommerceProxy.ProxyError[]) => {
                    Core.LogEvent(CommerceProxy.RetailLogger.updateWishListError, errors, Resources.String_253); // Sorry, something went wrong. We couldn't edit the wish list details. Please refresh the page and try again.
                    var errorMessages: string[] = ErrorHelper.getErrorMessages(errors);
                    this.showError(true);
                    LoadingOverlay.CloseLoadingDialog();
                });
        }
    }
}